### PR TITLE
Alphabetize gecko codes

### PR DESCRIPTION
**Pull request by jaylthegreat**
This adds a feature to sort the list of gecko codes alphabetically. A sort button was added to the GUI in the gecko cheat code menu for the Qt version of Dolphin. The codes are not permanently sorted, and will be returned to the natural ordering if you leave and re-open the menu.

### CODE REVIEW

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp [hunk 1]
```diff
@@ -220,10 +223,20 @@ void GeckoCodeWidget::SaveCodes()
   game_ini_local.Save(File::GetUserPath(D_GAMESETTINGS_IDX) + m_game_id + ".ini");
 }
 
+bool helpOrderCodes(Gecko::GeckoCode code1, Gecko::GeckoCode code2) { // to sort the codes alphabetically
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp]: This function isn't necessary and a lambda function should be passed to std::sort. i.e.

```cpp
std::sort(m_gecko_codes.begin(), m_gecko_codes.end(),
          [](const auto& lhs, const auto& rhs) { return lhs.name < rhs.name; });
```

File: Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp [hunk 2]
```diff
@@ -220,10 +223,20 @@ void GeckoCodeWidget::SaveCodes()
...
+  return (code1.name < code2.name);
+}
+
+void GeckoCodeWidget::SortCode(){
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp]: Brace goes on the next line

File: Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp [hunk 3]
```diff
@@ -220,10 +223,20 @@ void GeckoCodeWidget::SaveCodes()
...
+  std::sort(m_gecko_codes.begin(), m_gecko_codes.end(), helpOrderCodes);
+  GeckoCodeWidget::UpdateList();
+}
+
 void GeckoCodeWidget::UpdateList()
 {
   m_code_list->clear();
 
+  // std::sort(m_gecko_codes.begin(), m_gecko_codes.end(), helpOrderCodes);  // should sort the codes in alphabetical order
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp]: Commented code shouldn't be committed.

File: Source/Core/DolphinQt2/Config/GeckoCodeWidget.h [hunk 4]
```diff
@@ -46,6 +46,7 @@ class GeckoCodeWidget : public QWidget
   void RemoveCode();
   void DownloadCodes();
   void SaveCodes();
+  void SortCode();
```
Comment (lioncash) [inline Source/Core/DolphinQt2/Config/GeckoCodeWidget.h]: `SortCodesLexicographically()` would also indicate the sorting type being performed (i.e. `operator<` on two strings is defined as a general lexicographical comparison).

File: Source/Core/DolphinWX/Cheats/GeckoCodeDiag.cpp [hunk 5]
```diff
@@ -75,12 +75,16 @@ CodeConfigPanel::CodeConfigPanel(wxWindow* const parent) : wxPanel(parent)
   SetSizerAndFit(sizer_main);
 }
 
+bool helpOrderCodes(GeckoCode& code1, GeckoCode& code2) { // to sort the codes by name
```
Comment (lioncash) [inline Source/Core/DolphinWX/Cheats/GeckoCodeDiag.cpp]: Ditto here about using a lambda function

File: Source/Core/DolphinWX/Cheats/GeckoCodeDiag.cpp [hunk 6]
```diff
@@ -75,12 +75,16 @@ CodeConfigPanel::CodeConfigPanel(wxWindow* const parent) : wxPanel(parent)
...
+  return (code1.name < code2.name);
+}
+
 void CodeConfigPanel::UpdateCodeList(bool checkRunning)
 {
   // disable the button if it doesn't have an effect
   btn_download->Enable((!checkRunning || Core::IsRunning()) && !m_gameid.empty());
 
-  m_listbox_gcodes->Clear();
+  std::sort(m_gcodes.begin(), m_gcodes.end(), helpOrderCodes);  // should sort the codes in alphabetical order  m_listbox_gcodes->Clear();
```
Comment (lioncash) [inline Source/Core/DolphinWX/Cheats/GeckoCodeDiag.cpp]: Was commenting out the clear intended? If so, it should just be removed. Also comments generally go on top of the line, not off to the side in the codebase.

File: Source/Core/InputCommon/ControlReference/ControlReference.cpp [hunk 7]
```diff
@@ -105,8 +105,8 @@ ControlState InputReference::State(const ControlState ignore)
 //
 ControlState OutputReference::State(const ControlState state)
 {
-  if (m_parsed_expression && InputGateOn())
-    m_parsed_expression->SetValue(state);
+  if (m_parsed_expression)
```
Comment (lioncash) [inline Source/Core/InputCommon/ControlReference/ControlReference.cpp]: Why is this change included? It seems unrelated

File: Source/dolphin-emu.sln [hunk 8]
```diff
@@ -295,4 +295,7 @@ Global
 		{2C0D058E-DE35-4471-AD99-E68A2CAF9E18} = {87ADDFF9-5768-4DA2-A33B-2477593D6677}
 		{5BDF4B91-1491-4FB0-BC27-78E9A8E97DC3} = {87ADDFF9-5768-4DA2-A33B-2477593D6677}
 	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
```
Comment (lioncash) [inline Source/dolphin-emu.sln]: Ditto

File: Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp [hunk 9]
```diff
@@ -66,6 +66,7 @@ void GeckoCodeWidget::CreateWidgets()
   m_edit_code = new QPushButton(tr("&Edit Code..."));
   m_remove_code = new QPushButton(tr("&Remove Code"));
   m_download_codes = new QPushButton(tr("Download Codes"));
+  m_sort_code = new QPushButton(tr("Sort Codes"));
```
Comment (spycrab) [inline Source/Core/DolphinQt2/Config/GeckoCodeWidget.cpp]: ``m_sort_codes``

**Comment by leoetlino** [comment]
Superseded by #6699.
